### PR TITLE
Upgrade to Netty-4.1

### PR DESCRIPTION
**Pull request by chtyim**
The changes are already grouped into commits. There are two main changes:

1. Remove gson from the HttpResponder API
2. Switch the implementation to use netty-4.1

Other commits are just simple cleanup and moving the internal classes to an internal package.

### CODE REVIEW

File: src/main/java/co/cask/http/NettyHttpService.java [hunk 1]
```diff
@@ -272,23 +166,108 @@ public InetSocketAddress getBindAddress() {
...
     }
     LOG.debug("Stopped HTTP Service {} on address {}", serviceName, bindAddress);
   }
 
   /**
+   * Create {@link EventExecutorGroup} for executing handle methods.
+   *
+   * @param size size of threadPool
+   * @param keepAliveSecs  maximum time that excess idle threads will wait for new tasks before terminating
+   * @return instance of {@link EventExecutorGroup} or {@code null} if {@code size} is {@code <= 0}.
```
Comment (anwar6953) [inline src/main/java/co/cask/http/NettyHttpService.java:189]: I think the `@Nullable` is wrong. It doesn't seem to actually return null.

File: src/main/java/co/cask/http/RequestRouter.java [hunk 2]
```diff
@@ -16,132 +16,91 @@
...
     String exceptionMessage = "Exception caught in channel processing.";
-    Throwable cause = e.getCause();
     if (exceptionRaised.compareAndSet(false, true)) {
       if (methodInfo != null) {
         LOG.error(exceptionMessage, cause);
         methodInfo.sendError(HttpResponseStatus.INTERNAL_SERVER_ERROR, cause);
         methodInfo = null;
       } else {
-        ChannelFuture future = Channels.future(ctx.getChannel());
-        future.addListener(ChannelFutureListener.CLOSE);
```
Comment (anwar6953) [inline src/main/java/co/cask/http/RequestRouter.java]: why move the `addListener` call to the two cases below?

File: src/main/java/co/cask/http/internal/RequestRouter.java [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+      } else {
+        if (cause instanceof HandlerException) {
+          HttpResponse response = ((HandlerException) cause).createFailureResponse();
+          // trace logs for user errors, error logs for internal server errors
+          if (isUserError(response)) {
+            LOG.trace(exceptionMessage, cause);
+          } else {
+            LOG.error(exceptionMessage, cause);
+          }
+          ctx.writeAndFlush(response).addListener(ChannelFutureListener.CLOSE);
```
Comment (anwar6953) [inline src/main/java/co/cask/http/internal/RequestRouter.java:113]: I think this comment applies here: https://github.com/caskdata/netty-http/pull/46#discussion_r143300495

File: src/main/java/co/cask/http/NettyHttpService.java (hunk 1 above)
Comment (chtyim) [reply src/main/java/co/cask/http/NettyHttpService.java:189]: Yeah, there are some old annotations/docs hasn't been cleanup yet.

File: src/main/java/co/cask/http/RequestRouter.java (hunk 2 above)
Comment (chtyim) [reply src/main/java/co/cask/http/RequestRouter.java]: We don't need to create a new future. The `ctx.writeAndFlush` already returns a future for that operation.

**Review by anwar6953 — APPROVED** [review]

**Comment by chtyim** [comment]
squashed commits

**Pull request merged**